// @angular-eslint/eslint-plugin-template/dist/rules/use-track-by-function.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RULE_NAME = void 0;
const bundled_angular_compiler_1 = require("@angular-eslint/bundled-angular-compiler");
const utils_1 = require("@angular-eslint/utils");
const create_eslint_rule_1 = require("../utils/create-eslint-rule");
exports.RULE_NAME = 'use-track-by-function';
exports.default = (0, create_eslint_rule_1.createESLintRule)({
    name: exports.RULE_NAME,
    meta: {
        type: 'suggestion',
        docs: {
            description: 'Ensures trackBy function is used',
            recommended: false,
        },
        schema: [],
        messages: {
            useTrackByFunction: 'Missing trackBy function in ngFor directive',
        },
    },
    defaultOptions: [],
    create(context) {
        const parserServices = (0, utils_1.getTemplateParserServices)(context);
        return {
            'BoundAttribute.inputs[name="ngForOf"]'({ parent: { inputs }, sourceSpan, }) {
                if (inputs.some(isNgForTrackBy)) {
                    return;
                }
                const loc = parserServices.convertNodeSourceSpanToLoc(sourceSpan);
                context.report({
                    messageId: 'useTrackByFunction',
                    loc,
                });
            },
            'BoundAttribute.templateAttrs[name="ngForOf"]'({ parent: { templateAttrs }, }) {
                if (templateAttrs.some(isNgForTrackBy)) {
                    return;
                }
                const { start } = parserServices.convertNodeSourceSpanToLoc(templateAttrs[0].sourceSpan);
                const { end } = parserServices.convertNodeSourceSpanToLoc(templateAttrs[templateAttrs.length - 1].sourceSpan);
                const loc = {
                    start: Object.assign(Object.assign({}, start), { column: start.column - 1 }),
                    end: Object.assign(Object.assign({}, end), { column: end.column + 1 }),
                };
                context.report({
                    messageId: 'useTrackByFunction',
                    loc,
                });
            },
        };
    },
});
function isNgForTrackBy(attribute) {
    return (attribute instanceof bundled_angular_compiler_1.TmplAstBoundAttribute &&
        attribute.name === 'ngForTrackBy');
}
